### PR TITLE
Add System Catalog Views for Partitions Metadata

### PR DESCRIPTION
**Pull request by sumitj824**
### Description

This commit introduces support for the following system views:

1.  sys.partitions: To provide information about the individual partitions present in the database for both partitioned and non-partitioned indexes/tables.
2. sys.data_spaces: To provide information about the data spaces utilized by partition schemes.
3. sys.destination_data_spaces: To provide information about the destination data spaces associated with partition schemes.

Furthermore, this commit also updates the definition of the following existing views:

1. sys.indexes: Updates the definition to list HEAP entries for partitioned tables and updates the data_space_id column to the ID of the associated partition scheme.
2. sys.index_columns: Updates the definition to list HEAP entries for partitioned tables and entries for indexes of partitioned tables, which will be used to find the partitioning column.
3. sys.filegroups: Updates the datatype of the name column from SYSNAME to NVARCHAR(128).

These views, along with sys.partition_functions, sys.partition_schemes, sys.partition_parameters, and sys.partition_range_values which was added in 1ad41811a76292d7cdcc4402a080288118d1d233, provide comprehensive information about the partitioning configuration in the database.

Signed-off-by: Sumit Jaiswal [sumiji@amazon.com](mailto:sumiji@amazon.com)

### Issues Resolved

Task: BABEL-4970

### Test Scenarios Covered ###
* **Use case based -** Yes

* **Boundary conditions -** NA

* **Arbitrary inputs -** NA

* **Negative test cases -** NA

* **Minor version upgrade tests -** Yes

* **Major version upgrade tests -** Yes

* **Performance tests -** NA

* **Tooling impact -** NA

* **Client tests -** Yes

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/sql/sys_views.sql
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/sql/sys_views.sql:866]: Can we get the performance regression numbers for these changes?

File: contrib/babelfishpg_tsql/sql/sys_views.sql
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/sql/sys_views.sql:2617]: Need more explanation for this

File: contrib/babelfishpg_tsql/sql/sys_views.sql
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/sql/sys_views.sql]: Do we need sys.schemas here? Why cant we be more efficient/cleaner by using babelfish_namespace_ext and db_id

Comment (KushaalShroff) [reply contrib/babelfishpg_tsql/sql/sys_views.sql]: Same for all the usages of sys.schemas elsewhere, please re-check

File: contrib/babelfishpg_tsql/sql/sys_views.sql
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/sql/sys_views.sql]: WHERE ptbl.relkind = 'p', is there no other way to filter ctbl rather than joining with pg_class?

File: contrib/babelfishpg_tsql/sql/sys_views.sql
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/sql/sys_views.sql]: Same here, lets discuss here

**Review by KushaalShroff — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/sql/sys_views.sql
Comment (sumitj824) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:2617]: discussed it offline, it is need for DDL export to generate partition scheme with correct filegroups.

File: contrib/babelfishpg_tsql/sql/sys_views.sql
Comment (sumitj824) [reply contrib/babelfishpg_tsql/sql/sys_views.sql]: addressed everywhere.

File: contrib/babelfishpg_tsql/sql/sys_views.sql
Comment (sumitj824) [reply contrib/babelfishpg_tsql/sql/sys_views.sql]: removed the extra join on pg_class

File: contrib/babelfishpg_tsql/sql/sys_views.sql
Comment (sumitj824) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:866]: I will get the performance numbers for critical views separately.

Comment (KushaalShroff) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:866]: So shall we wait until we get it merged?

Comment (sumitj824) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:866]: No, it is not a blocker and can be done later.

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.2.0--4.3.0.sql
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.2.0--4.3.0.sql:11331]: I hope this will not leak any PG objects? If we do, did we can check with Alex if its fine since customers already might be using it as a workaround before this support for partitioning in TSQL?

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.2.0--4.3.0.sql
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.2.0--4.3.0.sql:11334]: Can we even do this in TSQL?
If not then we dont need cidx, right?

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.2.0--4.3.0.sql
Comment (sumitj824) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.2.0--4.3.0.sql:11331]: It will also list the partitions created it babelfish schemas from PG endpoint, I will check with Alex and block in separate PR depending on the reply.

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.2.0--4.3.0.sql
Comment (KushaalShroff) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.2.0--4.3.0.sql:11334]: As we spoke offline: Lets use imap only rather than cidx, since they are equivalent.

Comment (sumitj824) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.2.0--4.3.0.sql:11334]: I will evaluate the performance of sys.partitions along with sys.indexes and sys.index_column and accordingly will update the definition in separate PR if needed.

Comment (KushaalShroff) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.2.0--4.3.0.sql:11334]: Okay, then we can merge this provided that we followup on this quickly by raising a new PR - particularly the select * from pg_index in the CTE

**Review by KushaalShroff — APPROVED** [review]

**Pull request merged**
